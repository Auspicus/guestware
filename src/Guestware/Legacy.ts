import fetch from 'node-fetch'
import { DOMParser } from 'xmldom'
import { useNamespaces as XPathSelectWithNamespaces } from 'xpath'

import Client from './Client'
import Configuration from './Configuration'

import ReadGuestLoginRequest from './Soap/Method/ReadGuestLoginRequest'
import ReadGuestLoginGuestIDStringRequest from './Soap/Method/ReadGuestLoginGuestIDStringRequest'
import ReadGuestVisitRequest from './Soap/Method/ReadGuestVisitRequest'
import ReadGuestDetailTablesRequest from './Soap/Method/ReadGuestDetailTablesRequest'
import ReadGuestRewardTransactionAndDetailsByGuestIDRequest from './Soap/Method/ReadGuestRewardTransactionAndDetailsByGuestIDRequest'
import ReadGuestRewardBalanceRequest from './Soap/Method/ReadGuestRewardBalanceRequest'

import DatasetGuest from './Soap/Dataset/DatasetGuest'
import DatasetGuestRow from './Soap/Dataset/DatasetGuestRow'
import DiffgramRowAction from './Soap/DiffgramRowAction'
import UpdateGuestDetailTablesRequest from './Soap/Method/UpdateGuestDetailTablesRequest'

class Legacy {
  
  parser: DOMParser
  client: Client

  constructor(
    wsdl: string,
    appName: string,
    version: string,
    appId: string,
    username: string,
    password: string
  ) {
    const configuration = new Configuration({
      wsdl,
      appName,
      version,
      username,
      password
    })
    this.parser = new DOMParser()
    this.client = new Client({
      transport: fetch,
      configuration,
    })
  }

  async getGuestInformationByID(id: string): Promise<{
    id: string
    email?: string
    password?: string
    language?: string
    logActivity?: string
    disableLogin?: string
    created?: string
    createdBy?: string
    updated?: string
    updatedBy?: string
  }> {
    const xmlResponse = await this.client.send(new ReadGuestLoginRequest(id))
    return (
      this.formatResponse(this.parser.parseFromString(xmlResponse), {
        liTagName: 'GUEST_LOGIN',
        map: {
          id: 'GuestID',
          email: 'GuestLoginID',
          password: 'GuestLoginPassword',
          language: 'CultureID',
          logActivity: 'LogActivity',
          disableLogin: 'DisableLogin',
          created: 'EntryDate',
          createdBy: 'EntryBy',
          updated: 'LastEditDate',
          updatedBy: 'LastEditBy'
        }
      }).shift()
    )
  }

  async getGuestInformationByEmail(email: string): Promise<{
    id: string
    email?: string
    password?: string
    language?: string
    logActivity?: string
    disableLogin?: string
    created?: string
    createdBy?: string
    updated?: string
    updatedBy?: string
  }> {
    const xmlResponse = await this.client.send(new ReadGuestLoginGuestIDStringRequest(email))
    return (
      this.formatResponse(this.parser.parseFromString(xmlResponse), {
        liTagName: 'GUEST_LOGIN',
        map: {
          id: 'GuestID',
          email: 'GuestLoginID',
          password: 'GuestLoginPassword',
          language: 'CultureID',
          logActivity: 'LogActivity',
          disableLogin: 'DisableLogin',
          created: 'EntryDate',
          createdBy: 'EntryBy',
          updated: 'LastEditDate',
          updatedBy: 'LastEditBy'
        }
      }).shift()
    )
  }

  async getGuestDetails(id: string) {
    const xmlResponse = await this.client.send(new ReadGuestDetailTablesRequest(id))
    const parsed = this.parser.parseFromString(xmlResponse)
    return ({
      guest: this.formatResponse(parsed, {
        liTagName: 'GUEST',
        map: {
          id: 'GuestID',
          program: 'GstRecordType',
          type: 'ProfileType',
          source: 'Source',
          first_name: 'GivenName',
          last_name: 'Surname',
          created: 'EntryDate',
          createdBy: 'EntryBy',
          updated: 'LastEditDate',
          updatedBy: 'LastEditBy'
        }
      }),
      classes: this.formatResponse(parsed, {
        liTagName: 'GUEST_CLASS',
        map: {
          id: 'GuestID',
          classId: 'GuestClassID',
          className: 'ClassName',
          classReason: 'ClassReason',
          location: 'LocationID',
          start: 'StartDate',
          end: 'EndDate',
          created: 'EntryDate',
          createdBy: 'EntryBy',
          updated: 'LastEditDate',
          updatedBy: 'LastEditBy'
        }
      }),
      communications: this.formatResponse(parsed, {
        liTagName: 'GUEST_COMM_METHOD',
        map: {
          id: 'GuestID',
          type: 'GuestCommType',
          value: 'CommValue',
          format: 'FormatType',
          optOut: 'OptOut',
          preferred: 'PreferredFlag',
          created: 'EntryDate',
          createdBy: 'EntryBy',
          updated: 'LastEditDate',
          updatedBy: 'LastEditBy'
        }
      }),
      events: this.formatResponse(parsed, {
        liTagName: 'GUEST_EVENT',
        map: {
          id: 'GuestID',
          type: 'GuestEventType',
          order: 'DisplayOrder'
        }
      }),
      ids: this.formatResponse(parsed, {
        liTagName: 'GUEST_ISSUED_ID',
        map: {
          id: 'GuestID',
          logNumber: 'IssueLogNo',
          contactType: 'ContactType',
          associateName: 'AssociateName',
          created: 'EntryDate',
          createdBy: 'EntryBy',
          updated: 'LastEditDate',
          updatedBy: 'LastEditBy'
        }
      }),
      addresses: this.formatResponse(parsed, {
        liTagName: 'GUEST_ADDRESS',
        map: {
          id: 'GuestID',
          type: 'AddressType',
          string: 'CompleteAddress',
          country: 'CountryCode',
          postal: 'PostalCode',
          state: 'StateCode',
          city: 'City',
          address1: 'AddressLine1',
          address2: 'AddressLine2',
          undeliverable: 'Undeliverable',
          preferred: 'PreferredFlag',
          created: 'EntryDate',
          createdBy: 'EntryBy',
          updated: 'LastEditDate',
          updatedBy: 'LastEditBy'
        }
      }),
      interests: this.formatResponse(parsed, {
        liTagName: 'GUEST_INTERESTS',
        map: {
          id: 'GuestID',
          interest: 'Interest',
          commFormatType: 'CommFormatType',
          optOut: 'OptOut',
          created: 'EntryDate',
          createdBy: 'EntryBy',
          updated: 'LastEditDate',
          updatedBy: 'LastEditBy'
        }
      }),
      logins: this.formatResponse(parsed, {
        liTagName: 'GUEST_LOGIN',
        map: {
          id: 'GuestID',
          email: 'GuestLoginID',
          password: 'GuestLoginPassword',
          language: 'CultureID',
          logActivity: 'LogActivity',
          disableLogin: 'DisableLogin',
          created: 'EntryDate',
          createdBy: 'EntryBy',
          updated: 'LastEditDate',
          updatedBy: 'LastEditBy'
        }
      })
    })
  }

  async getGuestVisits(guestID: string, locationID?: string): Promise<object[]> {
    const xmlResponse = await this.client.send(new ReadGuestVisitRequest(guestID, locationID))
    return (
      this.formatResponse(this.parser.parseFromString(xmlResponse), {
        liTagName: 'GUEST_VISIT',
        map: {
          id: 'VisitLogNo',
          confirmation: 'ConfirmationNo',
          location: 'LocationID',
          status: 'VisitStatus',
          arrival: 'ArrivalDate',
          departure: 'DepartureDate',
          reservationDate: 'ReservationDate',
          reservationSource: 'ReservationSource',
          roomNumber: 'RoomNo',
          roomType: 'RoomType',
          roomRate: 'RoomRate',
          revenueRoom: 'VisitRoomRevenue',
          revenueFb: 'VisitFBRevenue',
          revenueMisc: 'VisitMiscRevenue',
          revenueUserDef1: 'UserDefRevenue1',
          revenueUserDef2: 'UserDefRevenue2',
          currency: 'CurrencyCode',
          marketCode: 'MarketCode',
          note: 'VisitNote',
          numberOfVisits: 'NumberOfVisits',
          numberOfNights: 'NumberOfNights',
          created: 'EntryDate',
          createdBy: 'EntryBy',
          updated: 'LastEditDate',
          updatedBy: 'LastEditBy',
          folioId: 'FolioID',
          visitRecordType: 'VisitRecordType'
        }
      })
    )
  }

  async getGuestRewards(guestID: string) {
    const xmlResponse = await this.client.send(new ReadGuestRewardTransactionAndDetailsByGuestIDRequest(guestID))

    return {
      parsed: this.parser.parseFromString(xmlResponse),
      raw: xmlResponse
    }
  }

  async getGuestRewardBalance(guestID: string) {
    const xmlResponse = await this.client.send(new ReadGuestRewardBalanceRequest(guestID))

    return {
      parsed: this.parser.parseFromString(xmlResponse),
      raw: xmlResponse
    }
  }

  async updateGuestDetails(details: {
    type: string,
    updated: boolean,
    properties: { [fieldName: string]: string }
  }[]): Promise<{ parsed: Document, raw: string }> {
    const dataset = this.detailMapToDataset(details)
    const xmlResponse = await this.client.send(new UpdateGuestDetailTablesRequest(dataset))

    return {
      parsed: this.parser.parseFromString(xmlResponse),
      raw: xmlResponse
    }
  }

  detailMapToDataset(details: {
    type: string,
    updated: boolean,
    properties: { [fieldName: string]: string }
  }[]): DatasetGuest {
    return new DatasetGuest(details.map(detail => {
      return new DatasetGuestRow(
        detail.type,
        detail.updated ? DiffgramRowAction.Modified : DiffgramRowAction.Inserted,
        detail.properties
      )
    }))
  }

  formatResponse(parsed: Document, options: {
    liTagName: string,
    map: { [name: string]: string }
  }) {
    const { liTagName, map } = options
    const result = []
    const XPathSelect = XPathSelectWithNamespaces({ dstGST: 'http://webservices.guestware.com/dstGST.xsd' })
    XPathSelect(`//dstGST:${liTagName}`, parsed).forEach(node => {
      const item = {}
      Object.keys(map).forEach(key => {
        item[key] = XPathSelect(`dstGST:${map[key]}/text()`, <Node> node, true).toString()
      })
      result.push(item)
    })
    return result
  }

}

export default Legacy