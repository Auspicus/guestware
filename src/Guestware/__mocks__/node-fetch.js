let mockResponses
const __setMockResponses = (responses) => mockResponses = responses
export { __setMockResponses }

const fetch = async () => Promise.resolve(mockResponses.shift())
export default fetch