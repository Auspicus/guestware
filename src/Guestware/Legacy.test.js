jest.mock('node-fetch')

import fs from 'fs'
import path from 'path'
import { DOMParser } from 'xmldom'

import Legacy from '../../lib/index.js'
import { __setMockResponses } from 'node-fetch'

const readMockData = async (name) => await new Promise(resolve => {
  fs.readFile(path.join(__dirname, './__data__', name), 'utf8', (err, data) => {
    if (err)
      throw err
    else
      resolve(data)
  })
})

describe('Legacy', () => {

  const parser = new DOMParser()
  const legacy = new Legacy(
    'wsdl',
    'appName',
    'version',
    'appId',
    'username',
    'password'
  )

  test('.formatResponse => converts Node to object based on map', async () => {
    expect.assertions(1)
    const xml = await readMockData('ReadGuestLogin.xml')
    const doc = parser.parseFromString(xml)
    const options = {
      liTagName: 'GUEST_LOGIN',
      map: {
        id: 'GuestID',
        email: 'GuestLoginID',
        password: 'GuestLoginPassword'
      }
    }
    expect(legacy.formatResponse(doc, options)).toMatchSnapshot()
  })

  test('.detailMapToDataset => converts detail map array to Dataset', async () => {
    expect(legacy.detailMapToDataset([
      {
        type: 'GUEST',
        updated: true,
        properties: {
          GuestID: '1234567890',
          GivenName: 'Bob',
          Surname: 'Dylan',
        }
      },
      {
        type: 'GUEST_INTERESTS',
        updated: false,
        properties: {
          GuestID: '1234567890',
          Interest: 'Cookies',
        }
      },
    ])).toMatchSnapshot()
  })

  test('.getGuestInformationByID => returns mapped response', async () => {
    expect.assertions(1)
    __setMockResponses([
      { ok: true, text: async () => await readMockData('ReadGuestLogin.xml') }
    ])
    await expect(legacy.getGuestInformationByID('1234567890')).resolves.toMatchSnapshot()
  })

  test('.getGuestInformationByEmail => returns mapped response', async () => {
    expect.assertions(1)
    __setMockResponses([
      { ok: true, text: async () => await readMockData('ReadGuestLoginGuestIDString.xml') }
    ])
    await expect(legacy.getGuestInformationByEmail('email@domain.com')).resolves.toMatchSnapshot()
  })

})