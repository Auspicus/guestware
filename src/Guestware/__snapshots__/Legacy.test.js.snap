// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Legacy .detailMapToDataset => converts detail map array to Dataset 1`] = `
DatasetGuest {
  "components": Array [
    DatasetGuestRow {
      "diff": "modified",
      "name": "GUEST",
      "properties": Object {
        "GivenName": "Bob",
        "GuestID": "1234567890",
        "Surname": "Dylan",
      },
      "updatedProperties": Object {},
    },
    DatasetGuestRow {
      "diff": "inserted",
      "name": "GUEST_INTERESTS",
      "properties": Object {
        "GuestID": "1234567890",
        "Interest": "Cookies",
      },
      "updatedProperties": Object {},
    },
  ],
}
`;

exports[`Legacy .formatResponse => converts Node to object based on map 1`] = `
Array [
  Object {
    "email": "{{Email}}",
    "id": "{{GuestID}}",
    "password": "{{Password}}",
  },
]
`;

exports[`Legacy .generateReadRequestBody => returns request to send in a read request 1`] = `"<soapenv:Envelope xmlns:soapenv=\\"http://schemas.xmlsoap.org/soap/envelope/\\" xmlns:web=\\"http://webservices.guestware.com/\\"><soapenv:Header><web:appId><web:UserName>username</web:UserName><web:PassWord>password</web:PassWord><web:ApplicationName>appName</web:ApplicationName><web:VersionNumber>version</web:VersionNumber></web:appId></soapenv:Header><soapenv:Body><web:ReadGuestDetailTables><web:parintGuestID>123456789</web:parintGuestID></web:ReadGuestDetailTables></soapenv:Body></soapenv:Envelope>"`;

exports[`Legacy .generateUpdateReadRequest => returns headers with appId 1`] = `"<soapenv:Envelope xmlns:soapenv=\\"http://schemas.xmlsoap.org/soap/envelope/\\" xmlns:web=\\"http://webservices.guestware.com/\\"><soapenv:Header><web:appId><web:UserName>username</web:UserName><web:PassWord>password</web:PassWord><web:ApplicationName>appName</web:ApplicationName><web:VersionNumber>version</web:VersionNumber></web:appId></soapenv:Header><soapenv:Body><web:ReadGuestDetailTables><web:parintGuestID>123456789</web:parintGuestID></web:ReadGuestDetailTables></soapenv:Body></soapenv:Envelope>"`;

exports[`Legacy .generateUpdateRequestBody => returns request to send in an update request 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><soap:Envelope xmlns:soap=\\"http://schemas.xmlsoap.org/soap/envelope/\\" xmlns:xsd=\\"http://www.w3.org/2001/XMLSchema\\" xmlns:xsi=\\"http://www.w3.org/2001/XMLSchema-instance\\"><soap:Header><appId xmlns=\\"http://webservices.guestware.com/\\"><UserName>username</UserName><PassWord>password</PassWord><ApplicationName>appName</ApplicationName><VersionNumber>version</VersionNumber></appId></soap:Header><soap:Body><UpdateGuestDetailTables xmlns=\\"http://webservices.guestware.com/\\"><pardstGuestDetailTables xmlns:msdata=\\"urn:schemas-microsoft-com:xml-msdata\\" msdata:SchemaSerializationMode=\\"ExcludeSchema\\"><xs:schema xmlns:xs=\\"http://www.w3.org/2001/XMLSchema\\" xmlns=\\"http://webservices.guestware.com/dstGST.xsd\\" xmlns:msprop=\\"urn:schemas-microsoft-com:xml-msprop\\" xmlns:mstns=\\"http://webservices.guestware.com/dstGST.xsd\\" id=\\"dstGST\\" targetNamespace=\\"http://webservices.guestware.com/dstGST.xsd\\" attributeFormDefault=\\"qualified\\" elementFormDefault=\\"qualified\\"><xs:element name=\\"dstGST\\" msdata:IsDataSet=\\"true\\" msdata:UseCurrentLocale=\\"true\\"><xs:complexType><xs:choice minOccurs=\\"0\\" maxOccurs=\\"unbounded\\"/></xs:complexType></xs:element></xs:schema><diffgr:diffgram xmlns:diffgr=\\"urn:schemas-microsoft-com:xml-diffgram-v1\\"><dstGST xmlns=\\"http://webservices.guestware.com/dstGST.xsd\\"><GUEST diffgr:id=\\"GUEST1\\" msdata:rowOrder=\\"0\\" diffgr:hasChanges=\\"inserted\\"><GuestID>0</GuestID><Surname>Last</Surname><GivenName>First</GivenName></GUEST></dstGST><diffgr:before></diffgr:before></diffgr:diffgram></pardstGuestDetailTables></UpdateGuestDetailTables></soap:Body></soap:Envelope>"`;

exports[`Legacy .getGuestInformationByEmail => returns mapped response 1`] = `
Object {
  "created": "2017-02-02T15:26:32.543-08:00",
  "createdBy": "Profile",
  "disableLogin": "false",
  "email": "{{Email}}",
  "id": "{{GuestID}}",
  "language": "{{Language}}",
  "logActivity": "false",
  "password": "{{Password}}",
  "updated": "2017-08-21T14:32:20.743-07:00",
  "updatedBy": "Profile",
}
`;

exports[`Legacy .getGuestInformationByID => returns mapped response 1`] = `
Object {
  "created": "2017-02-02T15:26:32.543-08:00",
  "createdBy": "Profile",
  "disableLogin": "false",
  "email": "{{Email}}",
  "id": "{{GuestID}}",
  "language": "{{Language}}",
  "logActivity": "false",
  "password": "{{Password}}",
  "updated": "2017-08-21T14:32:20.743-07:00",
  "updatedBy": "Profile",
}
`;
